src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,763 @@
+    this->got_len += to_read;
+
+    // Move pointer to the next record on the page
+    this->page_offset += MAXALIGN(this->tot_len);
+
+    // Rest of the record data is on the next page
+    while (this->got_len != this->record->xl_tot_len)
+    {
+stepReadBody:
+        if (readPage(this, input, stepReadBody))
+        {
+            return false;
+        }
+
+        if (this->current_header->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD)
+        {
+            // This record has been overwritten.
+            // Write to the output what we managed to read as is, skipping filtering.
+            return true;
+        }
+
+        if (!(this->current_header->xlp_info & XLP_FIRST_IS_CONTRECORD))
+        {
+            THROW_FMT(FormatError, "should be XLP_FIRST_IS_CONTRECORD");
+        }
+
+        if (this->current_header->xlp_rem_len == 0 ||
+            this->tot_len != (this->current_header->xlp_rem_len + this->got_len))
+        {
+            THROW_FMT(FormatError, "invalid contrecord length: expect: %zu, get %u",
+                      this->record->xl_tot_len - this->got_len, this->current_header->xlp_rem_len);
+        }
+
+        size_t to_write = Min(this->current_header->xlp_rem_len, this->walPageSize - this->page_offset);
+        memcpy(((char *) this->record) + this->got_len, this->page + this->page_offset, to_write);
+        this->page_offset += MAXALIGN(to_write);
+        this->got_len += to_write;
+    }
+    this->walInterface->validXLogRecord(this->record, this->heapPageSize);
+
+    this->currentStep = 0;
+
+    if (this->record->xl_rmid == RESOURCE_MANAGER_XLOG_ID && this->record->xl_info == SWITCH_XLOG)
+    {
+        this->is_switch_wal = true;
+    }
+    this->i++;
+    return true;
+}
+
+static void
+filterRecord(WalFilterState *const this)
+{
+    // In the case of overwrite contrecord, we do not need to try to filter it, since the record may not have a body at all.
+    if (this->got_len != this->record->xl_tot_len)
+    {
+        return;
+    }
+
+    RelFileNode *node;
+    // Pass through the records that are related to the system catalog or system databases (template1, template0 and postgres)
+    if (getRelFileNodeGPDB6(this->record, &node))
+    {
+        bool isPassTheFilter = false;
+
+        // isRelationNeeded can allocate memory on the first call. Therefore, we switch the context.
+        MEM_CONTEXT_OBJ_BEGIN(this)
+        isPassTheFilter = isRelationNeeded(node->dbNode, node->spcNode, node->relNode);
+        MEM_CONTEXT_OBJ_END();
+
+        if (isPassTheFilter)
+        {
+            return;
+        }
+
+        this->record->xl_rmid = RESOURCE_MANAGER_XLOG_ID;
+        this->record->xl_info = NOOP_XLOG;
+        uint32_t crc = crc32cInit();
+        crc = crc32cComp(crc, (unsigned char *) XLogRecGetData(this->record), this->record->xl_len);
+        crc = crc32cComp(crc, (unsigned char *) this->record, offsetof(XLogRecord, xl_crc));
+        crc = crc32cFinish(crc);
```

Comment:
Is it really necessary to calculate crc every time for the same nop record?

Reply:
We only change `xl_rmid` and `xl_info`. All other fields and the body of the record remain as is, so the checksum for each record will be its own.